catalyst/dl/runner/gan.py

```diff
@@ -61,18 +96,190 @@ def forward(self, batch):
+        for key, model in models.items():
+            assert isinstance(
+                model, torch.nn.Module
+            ), "model {key} must be torch.nn.Module type"
+        discriminator_assert_message = \
+            "models must have discriminator Module with {self.discriminator_key} key"  # noqa: E501
+        assert self.discriminator_key in models, discriminator_assert_message
+        assert isinstance(
+            models[self.discriminator_key], torch.nn.Module
+        ), discriminator_assert_message
+        generator_assert_message = \
+            "models must have generator Module with {self.generator_key} key"  # noqa: E501
+        assert self.generator_key in models, generator_assert_message
+        assert isinstance(
+            models[self.generator_key], torch.nn.Module
+        ), generator_assert_message
+
+    def prepare_state_params(
+        self, num_epochs: int, logdir: str, additional_state_params
+    ):
+        state_params = {
+            "num_epochs": num_epochs,
+            "logdir": logdir,
+            **self.TRAIN_PARAMS,
+        }
+        state_params.update(additional_state_params)
+        return state_params
+
+    @staticmethod
+    def prepare_experiment_params(
+        models,
+        criterion,
+        optimizers,
+        state_params,
+        callbacks,
+        datasets,
+        loaders,
+        train_stage_name="train",
+    ) -> Dict:
+        experiment_params = {
+            "models": {train_stage_name: models},
+            "criterions": {train_stage_name: criterion},
+            "optimizers": {train_stage_name: optimizers},
+            "stages": [train_stage_name],
+            "state_params": {train_stage_name: state_params},
+            "callbacks": {train_stage_name: callbacks},
+            "datasets": {train_stage_name: datasets},
+            "loaders": {train_stage_name: loaders},
+        }
+        return experiment_params
+
+    def prepare_callbacks(
+        self, callbacks, discriminator_phase_num, generator_phase_num
+    ) -> OrderedDict:
+        """
+            Catalyst can't take several optimizer callbacks for generator
+            and discriminator and process them raw
+            Therefore, phase manager, criterion callbacks and wrappers
+            will be created to manage training order
+        """
+        assert isinstance(
+            callbacks, dict
+        ), "callbacks should be of dict[str, Callback] type"
+        discriminator_phase = self.TRAIN_PARAMS["discriminator_train_phase"]
+        generator_phase = self.TRAIN_PARAMS["generator_train_phase"]
+        callbacks["phase_manager"] = PhaseManagerCallback(
+            train_phases=OrderedDict(
+                [
+                    (discriminator_phase, discriminator_phase_num),
+                    (generator_phase, generator_phase_num),
+                ]
+            ),
+            valid_mode="all",
+        )
+        callbacks["loss_g"] = PhaseBatchWrapperCallback(
```

Comment:
Isn't there a more beautiful way to determine this?
I can now propose these variants:
1. provide to `__init__` `generator_callbacks=[], discriminator_callbacks=[], shared_callbacks=[]` where each list contains `Callback`'s to be wrapped (and do not provide `callbacks=[]`)
2. provide to `__init__` 1)`callbacks=[...]` as it is right now and 2)`phase2callbacks=Dict[str, List[str]]` where the list contains the names(keys) of the activated callbacks

AND THIS SHOULD BE DONE IN EXPERIMENT CLASS, NOT IN RUNNER!